### PR TITLE
GHE-14 prepare for branch protection feature

### PR DESCRIPTION
**Pull request by fokusferit**
This PR prepares Zappr for the [new organizational branch protection feature](https://developer.github.com/v3/repos/branches/#update-required-status-checks-of-protected-branch).

It forces basically that removing a check for a protected branch triggers an error (403 Forbidden). There needs to be some adjustment in the DELETE Endpoint Error handling.

- [x] Test on dev
- [x] Test with prod stack

### CODE REVIEW

File: config/config.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ GITHUB_API_URL: https://api.github.com
 
 # Address of the web application host
 # that ZAPPR is going to run on.
-HOST_ADDR: http://localhost:8080
```
Comment (lotharschulz) [inline config/config.yaml]: that should stay localhost in the final version

File: server/service/GithubService.js [hunk 2]
```diff
@@ -385,6 +390,7 @@ export class GithubService {
                                  .replace('${repo}', repo)
                                  .replace('${branch}', branch)
     const settings = await this.getRequiredStatusChecks(user, repo, branch, accessToken)
+    debugger;
```
Comment (lotharschulz) [inline server/service/GithubService.js]: that is temporarily, correct?

File: package.json [hunk 3]
```diff
@@ -18,9 +18,11 @@
     "clean": "rm -rf dist",
     "all": "run-p client 'build-server -- -dw' server",
     "client": "webpack-dev-server -d -w --history-api-fallback --inline --no-info --config webpack.client.js --content-base dist/client",
-    "server": "nodemon dist/server/server.min.js --watch dist/server",
+    "server": "nodemon dist/server/server.min.js --watch  dist/server",
     "start": "node dist/server/server.min.js",
-    "scm": "scm-source > scm-source.json"
+    "scm": "scm-source > scm-source.json",
+    "migrate-webpack:server": "webpack-cli migrate webpack.server.js webpack.new.server.js",
```
Comment (fokusferit) [inline package.json]: should be removed.

File: server/service/GithubService.js [hunk 4]
```diff
@@ -61,15 +61,18 @@ export class GithubService {
   async fetchPath(method, path, payload, accessToken, headers = {}) {
     const options = this.getOptions(method, path, payload, accessToken, headers)
     const [response, body] = await request(options)
-    const {statusCode} = response || {}
-
+    const {statusCode} = response || {statusCode: 0}
```
Comment (fokusferit) [inline server/service/GithubService.js]: maybe not needed.

File: server/service/GithubService.js [hunk 5]
```diff
@@ -246,13 +249,15 @@ export class GithubService {
     // check if it's there already
     let hooks = await this.fetchPath('GET', path, null, accessToken)
     let existing = hooks.find(h => h.config.url === hook_url)
+    //TODO: AM I allowed?
```
Comment (fokusferit) [inline server/service/GithubService.js]: remove ToDo's

File: server/service/GithubService.js [hunk 6]
```diff
@@ -246,13 +249,15 @@ export class GithubService {
...
+    //TODO: AM I allowed? 
+    // MIssing CHECKS
     if (!!existing) {
       path += `/${existing.id}`
       if (payload.events.length) {
         await this.fetchPath('PATCH', path, payload, accessToken)
         debug(`${user}/${repo}: updated existing webhook ${existing.id}`)
       } else {
-        await this.fetchPath('DELETE', path, null, accessToken)
+        const deleteHook = await this.fetchPath('DELETE', path, null, accessToken)
```
Comment (fokusferit) [inline server/service/GithubService.js]: not needed.

**Comment by fokusferit** [comment]
👍

**Comment by hjacobs** [comment]
LGTM, any risk to merge it.

File: server/service/GithubService.js [hunk 7]
```diff
@@ -246,6 +249,7 @@ export class GithubService {
     // check if it's there already
     let hooks = await this.fetchPath('GET', path, null, accessToken)
     let existing = hooks.find(h => h.config.url === hook_url)
+    // MIssing CHECKS
```
Comment (hjacobs) [inline server/service/GithubService.js]: what kind of checks are missing?

Comment (fokusferit) [reply server/service/GithubService.js]: This was a past comment, when I had to go through the whole chaining of method calls before knowing exactly where the 403 Error should be handled.

**Comment by fokusferit** [comment]
I've added a test which checks that for a 403 error from github API the expected Error is thrown.
From my perspective it can be merged.
But I would ping @rashamalek and ask him to check the new feature on the GHE test (maybe).

**Comment by fokusferit** [comment]
👍

**Comment by fokusferit** [comment]
Pinging @lotharschulz

**Comment by lotharschulz** [comment]
@fokusferit: thanks for pinging. Your code changes LGTM. Let me propose some string literal changes. Will do that later today.

**Comment by lotharschulz** [comment]
@fokusferit : please check my recent changes. I hope these comments help to understand the changes better.

**Comment by lotharschulz** [comment]
👍

**Comment by rashamalek** [comment]
Today A performance (and potentially more) issue on Zappr-testbed was noticed, 
Clicking on "SYNC WITH GITHUB" button in zappr-testbed causes CPU load on GHE to reach nearly 100%

zapper version which is used in the test: 
"opensource/zappr-pr-test:5d41cd0-89"

[GHE 2.16](https://enterprise.github.com/releases/series/2.16) release notes states as a known issue in  : "Stricter REST API validation has been prematurely enabled. As a result, API requests that previously succeeded may be rejected with a 422 Unprocessable Entity response. (updated 2019-02-01) ->
" https://developer.github.com/changes/2018-09-25-stricter-validation-coming-soon-in-the-rest-api/

**Comment by fokusferit** [comment]
👍

**Comment by ChristianLohmann** [comment]
:+1:

**Comment by harti2006** [comment]
:+1:

**Pull request merged**